[torch.library] Add mechanism to add tags during define

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112053
* #111933
* #111948
* #111915
* __->__ #111912

We extend torch.library.Library.define and torch.library.define
with a tags argument.

Test Plan:
- new test

zou3519 committed:
[torch.library] Add mechanism to add tags during define

We extend torch.library.Library.define and torch.library.define
with a tags argument.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "[torch.library] Add mechanism to add tags during define"


We extend torch.library.Library.define and torch.library.define
with a tags argument. This makes them consistent with their C++ counterpart.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "[torch.library] Add mechanism to add tags during define"


We extend torch.library.Library.define and torch.library.define
with a tags argument.

Test Plan:
- new test

[ghstack-poisoned]

ezyang reviewed (approved):
It'll be convenient if it also takes list/single tag.

ezyang commented on torch/library.py:
```diff
@@ -71,13 +71,17 @@ def __init__(self, ns, kind, dispatch_key=""):
+            tags (optional): Tuple of torch.Tag to add to the operator. Adding
+                             torch.Tag may change the behavior of the operator,
+                             please throughly test the operator to ensure the
+                             Tag applies.
```
A typical end user isn't going to know what "test the operator so the tag applies" means.

zou3519 replied:
Thanks for the feedback, I'll link the gdoc.

zou3519 committed:
Update on "[torch.library] Add mechanism to add tags during define"


We extend torch.library.Library.define and torch.library.define
with a tags argument.

Test Plan:
- new test

[ghstack-poisoned]
